### PR TITLE
refactor: migrate evm

### PR DESCRIPTION
**Pull request by guanbinrui**
## Description

| Feature | Status | Assignee | 
| ------- | ------ | ----- |
| Gitcoin | Done   | @guanbinrui |
| Trader & Trending  | \-     | @nuanyang233 |
| Popup  | \-     | @nuanyang233 |
| Dashboard Asset | \- | @NikoManiac |
| Web3 Tab & Multichains | \- | @UncleBill |
| MaskBox & Multichains | Done | @UncleBill |
| Tips | Done | @UncleBill |
| ITO | \- | @zhouhanseng |
| RedPacket | \- | @zhouhanseng |
| Snapshot | Done | @zhouhanseng |
| Transak | Done | @guanbinrui |
| Furucombo | Done | @guanbinrui |
| vCent | N/A | \- |
| UnlockProtocol | N/A | \- |
| Pools | N/A | \- |
| DAO | N/A | \- |
| ArtBlock | \- | @Randolph314 |
| Findtrueman | \- | @BillySh1 |
| dHedge | Done  | @yanzhihong23 |
| CryptoartAI | Done | @NikoManiac |
| Collectible | \- | @NikoManiac |
| Avatar | \- | @lelenei |
| GoodGhosting | \- | @yanzhihong23 |
| Savings | \- | @BillySh1 |
| Pets | \- | @Randolph314 |
| NextID | \- | @yanzhihong23 |

## Type of change

- [ ] Documentation
- [x] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

### CODE REVIEW

File: packages/mask/src/extension/popups/pages/Wallet/ContractInteraction/index.tsx [hunk 1]
```diff
@@ -26,11 +26,11 @@ import { isGreaterThan, leftShift, pow10 } from '@masknet/web3-shared-base'
 import { unreachable } from '@dimensiondev/kit'
 import { useI18N } from '../../../../../utils'
 import { WalletRPC } from '../../../../../plugins/Wallet/messages'
-import Services from '../../../../service'
 import { useNativeTokenPrice, useTokenPrice } from '../../../../../plugins/Wallet/hooks/useTokenPrice'
 import { LoadingPlaceholder } from '../../../components/LoadingPlaceholder'
 import { NetworkPluginID, useReverseAddress, useWeb3State } from '@masknet/plugin-infra'
 import { CopyIconButton } from '../../../components/CopyIconButton'
+import { EVM_RPC } from '@masknet/plugin-evm/src/messages'
```
Comment (Jack-Works) [inline packages/mask/src/extension/popups/pages/Wallet/ContractInteraction/index.tsx]: you should not import from `src` or `dist` folder.

File: packages/plugins/EVM/package.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    "@masknet/web3-contracts": "workspace:*",
+    "@masknet/web3-providers": "workspace:*",
+    "@masknet/web3-shared-base": "workspace:*",
+    "@masknet/web3-shared-evm": "workspace:*",
+    "@metamask/eth-sig-util": "^4.0.0",
+    "bignumber.js": "^9.0.2",
+    "ethereumjs-util": "^7.1.0",
+    "react-use": "^17.3.2",
+    "wallet.ts": "^1.0.1",
+    "web3": "1.5.2",
```
Comment (Jack-Works) [inline packages/plugins/EVM/package.json]: 1.7.1

File: packages/web3-providers/src/mask/prepare.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+if (typeof self !== 'undefined') {
```
Comment (Jack-Works) [inline packages/web3-providers/src/mask/prepare.ts]: hmm what are you doing? why you need a worker in this package?

File: packages/shared-base/src/utils/subscription.ts [hunk 4]
```diff
@@ -73,6 +73,18 @@ export function mapSubscription<T, Q>(sub: Subscription<T>, mapper: (val: T) =>
     }
 }
 
+export function mergeSubscription<T>(...subs: Subscription<unknown>[]) {
```
Comment (Jack-Works) [inline packages/shared-base/src/utils/subscription.ts]: This is not type-safe. Please change to the correct type signature.

File: packages/shared-base/src/ondemand-worker/index.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
+/**
```
Comment (Jack-Works) [inline packages/shared-base/src/ondemand-worker/index.ts]: You should delete another copy of this file in this project.

File: packages/shared-base/src/Site/index.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+function getEnhanceableSiteType() {
+    for (const [type, regexp] of Object.entries(matchEnhanceableSiteHost)) {
+        if (host.match(regexp)) return type as EnhanceableSite
+        continue
+    }
+    return
+}
+
+function getExtensionSiteType() {
```
Comment (Jack-Works) [inline packages/shared-base/src/Site/index.ts]: Please try `getEnvironment` exported from the `'@dimensiondev/holoflows-kit'` package.

File: packages/shared-base/src/Settings/index.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import { defer } from '@dimensiondev/kit'
```
Comment (Jack-Works) [inline packages/shared-base/src/Settings/index.ts]: Please do not add new Settings API based on ValueRef.

Use `packages/shared-base/src/kv-storage`. You can find an example at `packages/mask/shared/kv-storage.ts`

File: packages/plugin-infra/src/types.ts [hunk 8]
```diff
@@ -1,11 +1,13 @@
...
 import type { Option, Result } from 'ts-results'
+import type { Subscription } from 'use-subscription'
+import type { JsonRpcPayload, JsonRpcResponse } from 'web3-core-helpers'
 import type { TypedMessage } from '@masknet/typed-message'
-import type { ScopedStorage, ProfileIdentifier, PersonaIdentifier } from '@masknet/shared-base'
+import type { ScopedStorage, ProfileIdentifier, PersonaIdentifier, PopupRoutes } from '@masknet/shared-base'
 import type { Emitter } from '@servie/events'
 import type { Web3Plugin } from './web3-types'
-import type { Subscription } from 'use-subscription'
+import type { EthereumTransactionConfig } from '../../web3-providers/node_modules/@masknet/web3-shared-evm'
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts]: Please do not import type like this.

File: packages/plugin-infra/src/types.ts [hunk 9]
```diff
@@ -61,15 +63,67 @@ export declare namespace Plugin {
...
+        /** Get the primary wallet */
+        walletPrimary: Subscription<Web3Plugin.Wallet | null>
+
+        /** A lightweight K/V storage used to store some simple data. */
         createKVStorage<T extends object>(type: 'memory' | 'persistent', defaultValues: T): ScopedStorage<T>
+
+        /** Native platform type */
+        nativeType?: 'Android' | 'iOS'
+        /** Native API supported */
+        hasNativeAPI: boolean
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts]: You do not need both. Only provide one.

File: packages/plugin-infra/src/types.ts [hunk 10]
```diff
@@ -61,15 +63,67 @@ export declare namespace Plugin {
...
+        /** iOS Ethereum send request */
+        nativeSend?: (payload: JsonRpcPayload) => Promise<JsonRpcResponse>
+        /** Android Ethereum send request */
+        nativeSendJsonString?: (message: string) => Promise<string>
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts]: Why expose this to the plugin?

File: packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx [hunk 11]
```diff
@@ -27,7 +27,7 @@ import { toAsset } from '../helpers'
 import { PluginTraderMessages } from '../../Trader/messages'
 import { Trans } from 'react-i18next'
 import getUnixTime from 'date-fns/getUnixTime'
-import type { useAsset } from '../../EVM/hooks'
+import type { useAsset } from '@masknet/plugin-evm/src/hooks'
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Collectible/SNSAdaptor/MakeOfferDialog.tsx]: Please do not import from `src/`

File: packages/mask/src/plugins/Collectible/apis/openseaFromSDK.ts [hunk 12]
```diff
@@ -1,12 +1,12 @@
 import { OpenSeaPort } from 'opensea-js'
 import { ChainId, createExternalProvider } from '@masknet/web3-shared-evm'
-import { request } from '../../../extension/background-script/EthereumService'
 import { resolveOpenSeaNetwork } from '../pipes'
 import { OpenSeaAPI_Key, ReferrerAddress } from '../constants'
+import { EVM_RPC } from '@masknet/plugin-evm/src/messages'
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Collectible/apis/openseaFromSDK.ts]: Please do not import from `src/`

**Review by Jack-Works — CHANGES_REQUESTED** [review]
pnpm:

>  WARN  There are cyclic workspace dependencies

TypeScript:

```
error TS6202: Project references may not form a circular graph. Cycle detected: tsconfig.json
packages/mask/tsconfig.json
packages/mask/src/tsconfig.json
packages/shared/tsconfig.json
packages/web3-shared/evm/tsconfig.json
packages/plugin-infra/tsconfig.json
packages/web3-providers/tsconfig.json
```

**Comment by Jack-Works** [comment]
image

This PR is horrifying big, is it possible to split it into multiple PRs?

File: packages/plugin-infra/src/web3-state/Provider.ts [hunk 13]
```diff
@@ -135,13 +135,13 @@ export class ProviderState<
...
             })
+        }
 
-            if (needToUpdateProviderType) {
-                await this.storage.providers.setValue({
-                    ...this.storage.providers.value,
-                    [siteType]: providerType,
-                })
-            }
+        if (needToUpdateProviderType) {
```
Comment (septs) [inline packages/plugin-infra/src/web3-state/Provider.ts]:
```suggestion
        if (!needToUpdateProviderType) return
```

**Pull request merged**